fiaas_deploy_daemon/crd/apiextensionsv1_crd_watcher.py

```diff
@@ -0,0 +1,220 @@
+# coding: utf-8
+
+# Copyright 2017-2019 The FIAAS Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import logging
+
+from k8s.base import WatchEvent
+from k8s.client import NotFound
+from k8s.models.common import ObjectMeta
+from k8s.models.apiextensions_v1_custom_resource_definition import CustomResourceConversion,\
+    CustomResourceDefinitionNames, CustomResourceDefinitionSpec, CustomResourceDefinition,\
+    CustomResourceDefinitionVersion, CustomResourceValidation, JSONSchemaProps
+from k8s.watcher import Watcher
+from yaml import YAMLError
+
+from .status import create_name
+from .types import FiaasApplication, FiaasApplicationStatus
+from ..base_thread import DaemonThread
+from ..deployer import DeployerEvent
+from ..log_extras import set_extras
+from ..retry import retry_on_upsert_conflict
+from ..specs.factory import InvalidConfiguration
+
+LOG = logging.getLogger(__name__)
+
+
+class ApiextensionsV1CrdWatcher(DaemonThread):
+    def __init__(self, spec_factory, deploy_queue, config, lifecycle):
+        super(ApiextensionsV1CrdWatcher, self).__init__()
+        self._spec_factory = spec_factory
+        self._deploy_queue = deploy_queue
+        self._watcher = Watcher(FiaasApplication)
+        self._lifecycle = lifecycle
+        self.namespace = config.namespace
+        self.enable_deprecated_multi_namespace_support = config.enable_deprecated_multi_namespace_support
+
+    def __call__(self):
+        while True:
+            if self.enable_deprecated_multi_namespace_support:
+                self._watch(namespace=None)
+            else:
+                self._watch(namespace=self.namespace)
+
+    def _watch(self, namespace):
+        try:
+            for event in self._watcher.watch(namespace=namespace):
+                self._handle_watch_event(event)
+        except NotFound:
+            self.create_custom_resource_definitions()
+        except Exception:
+            LOG.exception("Error while watching for changes on FiaasApplications")
+
+    @classmethod
+    def create_custom_resource_definitions(cls):
+        object_with_unknown_fields = {"type": "object", "x-kubernetes-preserve-unknown-fields": True}
+        application_schema_properties = {
```

Comment:
I tried testing this a bit locally now, and I think the `spec` top level object is missing here, i.e. all of these fields should be under `spec`

Reply:
I indeed missed that.